Can't call macro defined as `#[proc_macro_hack(support_nested)]` from output of macro  defined as `#[proc_macro_hack(support_nested)]`

Consider having a macro exported as `#[proc_macro_hack(support_nested)]` which inserts macro defined as `#[proc_macro_hack(support_nested)]`. We will get following error:
```
error[E0659]: `proc_macro_call_0` is ambiguous (macro-expanded name vs less macro-expanded name from outer scope during import/macro resolution)
```
This situation may occur if we create an async macro using `#[proc_macro_hack(support_nested)]`, which internally uses `join!` or `select!` from https://github.com/rust-lang-nursery/futures-rs. 

I didn't find any possibility to import `join` function declared in https://github.com/rust-lang-nursery/futures-rs/tree/master/futures-join-macro directly in order to process token stream during code generation in custom macro. Appreciate any help.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,7 +14,7 @@ proc-macro = true
 [dependencies]
 proc-macro2 = "1.0"
 quote = "1.0"
-syn = "1.0"
+syn = "1.0.5"
 
 [dev-dependencies]
 demo-hack = { version = "0.0.5", path = "demo-hack" }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -136,7 +136,7 @@ use std::fmt::Write;
 use proc_macro2::{Span, TokenStream, TokenTree};
 use quote::{format_ident, quote, ToTokens};
 use syn::parse::{Parse, ParseStream, Result};
-use syn::{braced, bracketed, parenthesized, parse_macro_input, token, Ident, Token};
+use syn::{braced, bracketed, parenthesized, parse_macro_input, token, Ident, LitInt, Token};
 
 type Visibility = Option<Token![pub]>;
 
@@ -274,11 +274,13 @@ pub fn proc_macro_hack(
 mod kw {
     syn::custom_keyword!(derive);
     syn::custom_keyword!(fake_call_site);
+    syn::custom_keyword!(internal_macro_calls);
     syn::custom_keyword!(support_nested);
 }
 
 struct ExportArgs {
     support_nested: bool,
+    internal_macro_calls: u16,
     fake_call_site: bool,
 }
 
@@ -286,6 +288,7 @@ impl Parse for ExportArgs {
     fn parse(input: ParseStream) -> Result<Self> {
         let mut args = ExportArgs {
             support_nested: false,
+            internal_macro_calls: 0,
             fake_call_site: false,
         };
 
@@ -294,6 +297,11 @@ impl Parse for ExportArgs {
             if ahead.peek(kw::support_nested) {
                 input.parse::<kw::support_nested>()?;
                 args.support_nested = true;
+            } else if ahead.peek(kw::internal_macro_calls) {
+                input.parse::<kw::internal_macro_calls>()?;
+                input.parse::<Token![=]>()?;
+                let calls = input.parse::<LitInt>()?.base10_parse()?;
+                args.internal_macro_calls = calls;
             } else if ahead.peek(kw::fake_call_site) {
                 input.parse::<kw::fake_call_site>()?;
                 args.fake_call_site = true;
@@ -414,7 +422,11 @@ fn expand_export(export: Export, args: ExportArgs) -> TokenStream {
     };
     let crate_prefix = vis.map(|_| quote!($crate::));
     let enum_variant = if args.support_nested {
-        quote!(Nested)
+        if args.internal_macro_calls == 0 {
+            quote!(Nested)
+        } else {
+            format_ident!("Nested{}", args.internal_macro_calls).to_token_stream()
+        }
     } else {
         quote!(Value)
     };
@@ -438,8 +450,9 @@ fn expand_export(export: Export, args: ExportArgs) -> TokenStream {
             };
 
             let proc_macro_call = if args.support_nested {
+                let extra_bangs = (0..args.internal_macro_calls).map(|_| quote!(!));
                 quote! {
-                    #crate_prefix #dispatch! { ($($proc_macro)*) }
+                    #crate_prefix #dispatch! { ($($proc_macro)*) #(#extra_bangs)* }
                 }
             } else {
                 quote! {
@@ -524,7 +537,8 @@ fn expand_define(define: Define) -> TokenStream {
                 _ => unimplemented!(),
             };
             let variant = braces.next().unwrap(); // `Value` or `Nested`
-            let support_nested = variant.to_string() == "Nested";
+            let varname = variant.to_string();
+            let support_nested = varname.starts_with("Nested");
             braces.next().unwrap(); // `=`
 
             let mut parens = match braces.next().unwrap() {
@@ -572,7 +586,12 @@ fn expand_define(define: Define) -> TokenStream {
                 #dummy::TokenTree::Ident(
                     #dummy::Ident::new(
                         &if support_nested {
-                            format!("proc_macro_call_{}", count_bangs(inner))
+                            let extra_bangs = if varname == "Nested" {
+                                0
+                            } else {
+                                varname["Nested".len()..].parse().unwrap()
+                            };
+                            format!("proc_macro_call_{}", extra_bangs + count_bangs(inner))
                         } else {
                             String::from("proc_macro_call")
                         },

